### PR TITLE
Fix correctness issue with DISTINCT grouping sets

### PR DESCRIPTION
**Pull request by martint**
Fixes https://github.com/prestodb/presto/issues/11120.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -1443,19 +1445,19 @@ private void checkGroupingSetsCount(GroupBy node)
             int crossProduct = 1;
             for (GroupingElement element : node.getGroupingElements()) {
                 try {
-                    int product;
+                    int product = 1;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: can we handle all supported cases explicitly (or via a visitor)?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -1476,6 +1478,18 @@ else if (element instanceof Rollup) {
 
             if (node.getGroupBy().isPresent()) {
                 checkGroupingSetsCount(node.getGroupBy().get());
+
+                // Only SimpleGroupBy allows complex expressions (a deviation from the SQL specification)
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: move this directly above `.filter(element -> !(element instanceof SimpleGroupBy))`

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 3]
```diff
@@ -932,6 +933,30 @@ public void testGroupByEmpty()
         assertFails(MUST_BE_AGGREGATE_OR_GROUP_BY, "SELECT a FROM t1 GROUP BY ()");
     }
 
+    @Test
+    public void testComplexExpressionInGroupingSet()
+    {
+        assertFails(MUST_BE_COLUMN_REFERENCE,
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java]: place first arg on a new line

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/GroupingElement.java [hunk 4]
```diff
@@ -25,6 +25,8 @@ protected GroupingElement(Optional<NodeLocation> location)
         super(location);
     }
 
+    public abstract List<Expression> getColumns();
```
Comment (findepi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/GroupingElement.java]: i am not sure this is the best name. `groupingColumns`? keys?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SimpleGroupBy.java [hunk 5]
```diff
@@ -45,7 +45,7 @@ private SimpleGroupBy(Optional<NodeLocation> location, List<Expression> simpleGr
         this.columns = ImmutableList.copyOf(requireNonNull(simpleGroupByExpressions, "simpleGroupByExpressions is null"));
     }
 
-    public List<Expression> getColumnExpressions()
+    public List<Expression> getColumns()
```
Comment (findepi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SimpleGroupBy.java]: `@Override` ?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java [hunk 6]
```diff
@@ -125,12 +125,15 @@ static long calculateGrouping(long groupId, List<Integer> columns, List<List<Int
...
-            if (index != -1) {
+
+        for (int index = 0; index < columns.size(); index++) {
+            int column = columns.get(index);
+
+            if (groupingSet.contains(column)) {
                 // Leftmost argument to grouping() (i.e. when index = 0) corresponds to
                 // the most significant bit in the result. That is why we shift 1L starting
                 // from the columns.size() - 1 bit index.
+
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java]: unnecessary change

File: presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java [hunk 7]
```diff
@@ -125,12 +125,15 @@ static long calculateGrouping(long groupId, List<Integer> columns, List<List<Int
...
 
         List<Integer> groupingSet = groupingSetDescriptors.get(toIntExact(groupId));
-        for (Integer groupingColumn : groupingSet) {
-            int index = columns.indexOf(groupingColumn);
-            if (index != -1) {
+
+        for (int index = 0; index < columns.size(); index++) {
+            int column = columns.get(index);
+
+            if (groupingSet.contains(column)) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java]: `groupingSet` is not a set, so a problem for (absurdly?) long `grouping(..)` list

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 8]
```diff
@@ -698,6 +714,46 @@ public TableHandle getTarget()
...
+        public GroupingSetAnalysis(
+                List<Set<FieldId>> cubes,
+                List<List<FieldId>> rollups,
+                List<List<Set<FieldId>>> ordinarySets,
+                List<Expression> complexExpressions)
+        {
+            this.cubes = cubes;
+            this.rollups = rollups;
+            this.ordinarySets = ordinarySets;
+            this.complexExpressions = complexExpressions;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: `ImmutableList.copyOf` each of them?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 9]
```diff
@@ -1472,107 +1473,100 @@ else if (element instanceof GroupingSets) {
...
-                    throw new SemanticException(TYPE_MISMATCH, node, "%s is not comparable, and therefore cannot be used in GROUP BY", type);
-                }
+                analysis.setGroupByExpressions(node, expressions);
+                analysis.setGroupingSets(node, new Analysis.GroupingSetAnalysis(cubes.build(), rollups.build(), sets.build(), complexExpressions.build()));
 
-                groupingColumnsBuilder.add(groupByExpression);
+                return expressions;
             }
-            return groupingColumnsBuilder.build();
+            else if (hasAggregates(node)) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: redundant else

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/GroupingElement.java (hunk 4 above)
Comment (martint) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/GroupingElement.java]: Well, it's the columns listed in the grouping element, so kind of obvious, no?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java (hunk 7 above)
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java]: Aha! This commit was meant to appear last, after groupingSet is made a set in "compute grouping sets during planning"

**Comment by martint** [comment]
@findepi, comments addressed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 10]
```diff
@@ -567,17 +592,64 @@ private PlanBuilder aggregate(PlanBuilder subPlan, QuerySpecification node)
...
+    private List<Set<FieldId>> enumerateGroupingSets(Analysis.GroupingSetAnalysis groupingSetAnalysis)
+    {
+        List<List<Set<FieldId>>> partialSets = new ArrayList<>();
+
+        for (Set<FieldId> cube : groupingSetAnalysis.getCubes()) {
+            partialSets.add(ImmutableList.copyOf(Sets.powerSet(cube)));
+        }
+
+        for (List<FieldId> rollup : groupingSetAnalysis.getRollups()) {
+            List<Set<FieldId>> sets = IntStream.range(0, rollup.size() + 1)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: `rangeClosed`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 11]
```diff
@@ -44,7 +42,9 @@
 {
     private final PlanNode source;
     private final Map<Symbol, Aggregation> aggregations;
-    private final List<List<Symbol>> groupingSets;
+    private final List<Symbol> groupingKeys;
+    private final int groupingSetCount;
+    private final Set<Integer> globalGroupingSets;
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: It's the `()` grouping set (aggregates globally)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 12]
```diff
@@ -698,6 +714,46 @@ public TableHandle getTarget()
         }
     }
 
+    public static class GroupingSetAnalysis
+    {
+        private final List<Set<FieldId>> cubes;
+        private final List<List<FieldId>> rollups;
+        private final List<List<Set<FieldId>>> ordinarySets;
+        private final List<Expression> complexExpressions;
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: I think that's redundant. These are complex expressions in the context of grouping sets analysis, so it's implied they come from the group by clause.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 13]
```diff
@@ -473,34 +472,60 @@ private PlanBuilder aggregate(PlanBuilder subPlan, QuerySpecification node)
...
+            groupingSetMappings.put(output, input);
+        }
+
+        // This tracks the grouping sets before complex expressions are considered (see comments below)
+        // It's also used to compute the descriptors needed to implement grouping()
+        List<Set<FieldId>> columnOnlyGroupingSets = ImmutableList.of(ImmutableSet.of());
+        List<List<Symbol>> groupingSets = ImmutableList.of(ImmutableList.of());
+
+        if (node.getGroupBy().isPresent()) {
+            // For the purpose of "distinct", we need to canonicalize column references that may have varying
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: > should not we canonicalize always, not only when node.getGroupBy().isPresent()?

This refers to canonicalizing the expressions that appear in the group by clause due to the semantics of DISTINCT with multiple grouping sets.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 14]
```diff
@@ -532,11 +534,27 @@ private PlanBuilder aggregate(PlanBuilder subPlan, QuerySpecification node)
         }
         Map<Symbol, Aggregation> aggregations = aggregationsBuilder.build();
 
+        Set<Integer> globalGroupingSets = new LinkedHashSet<>();
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: No, I need the indices here, not the actual grouping sets. But I can use an ImmutableSet.Builder instead.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 15]
```diff
@@ -236,7 +261,59 @@ public boolean hasSingleNodeExecutionPreference(FunctionRegistry functionRegistr
...
+            globalGroupingSets = ImmutableSet.of(0);
+        }
+        else {
+            globalGroupingSets = ImmutableSet.of();
+        }
+
+        return new GroupingSetDescriptor(groupingKeys, 1, globalGroupingSets);
+    }
+
+    public static GroupingSetDescriptor multipleGroupingSets(List<Symbol> groupingKeys, int groupingSetCount, Set<Integer> globalGroupingSets)
```
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: I see why you would prefer to be explicit about `singleGroupingSet`, but I might call this `groupingSets` because you can consider that to be the 'common' case.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 16]
```diff
@@ -236,7 +261,59 @@ public boolean hasSingleNodeExecutionPreference(FunctionRegistry functionRegistr
...
+    {
+        private final List<Symbol> groupingKeys;
+        private final int groupingSetCount;
+        private final Set<Integer> globalGroupingSets;
+
+        private GroupingSetDescriptor(List<Symbol> groupingKeys, int groupingSetCount, Set<Integer> globalGroupingSets)
+        {
+            this.groupingKeys = groupingKeys;
+            this.groupingSetCount = groupingSetCount;
+            this.globalGroupingSets = globalGroupingSets;
```
[the remark this answers is not in the document]
Comment (raghavsethi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: Yes, I think you might have missed requireNonNull/copyOf in the fixup.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/AggregationMatcher.java [hunk 17]
```diff
@@ -67,7 +67,15 @@ public MatchResult detailMatches(PlanNode node, StatsProvider stats, Session ses
             return NO_MATCH;
         }
 
-        if (groupingSets.size() != aggregationNode.getGroupingSets().size()) {
+        if (!matches(groupingSets.getGroupingKeys(), aggregationNode.getGroupingKeys(), symbolAliases)) {
```
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/AggregationMatcher.java]: These conditions feel pretty close to an equality check on GroupingSetsDescriptor

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java [hunk 18]
```diff
@@ -270,24 +270,20 @@ public AggregationBuilder addAggregation(Symbol output, Aggregation aggregation)
...
-        public AggregationBuilder groupingSets(List<List<Symbol>> groupingSets)
-        {
-            checkState(this.groupingSets.isEmpty(), "groupingSets already defined");
-            this.groupingSets.addAll(groupingSets);
+            groupingSets(AggregationNode.singleGroupingSet(ImmutableList.of()));
             return this;
         }
 
-        public AggregationBuilder addGroupingSet(Symbol... symbols)
+        public AggregationBuilder singleGroupingSet(Symbol... symbols)
```
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: Not a huge fan of builder methods without add, also the function name could be confusing because of conflict with util methods in AggregatioNode

Comment (martint) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/PlanBuilder.java]: It's not additive, though. You couldn't call `addSingleGroupingSet` multiple times and expect it to work.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java [hunk 19]
```diff
@@ -118,12 +137,12 @@ public boolean hasDefaultOutput()
 
     public boolean hasEmptyGroupingSet()
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/AggregationNode.java]: Will do as a follow-up PR. I didn't want to cause lots of unrelated changes everywhere that references this method.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/AggregationMatcher.java (hunk 17 above)
Comment (martint) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/AggregationMatcher.java]: Yeah, but it needs access to symbolAliases.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 20]
```diff
@@ -1476,6 +1484,17 @@ else if (element instanceof Rollup) {
...
             if (node.getGroupBy().isPresent()) {
                 checkGroupingSetsCount(node.getGroupBy().get());
+
+                node.getGroupBy().get().getGroupingElements().stream()
+                        .filter(element -> !(element instanceof SimpleGroupBy)) // Only SimpleGroupBy allows complex expressions (a deviation from the SQL specification)
+                        .flatMap(element -> element.getColumns().stream())
+                        .forEach(column -> {
+                            analyzeExpression(column, scope);
+                            if (!analysis.getColumnReferences().contains(NodeRef.of(column))) {
+                                throw new SemanticException(SemanticErrorCode.MUST_BE_COLUMN_REFERENCE, column, "GROUP BY expression must be a column reference: %s", column);
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: From a users perspective, there's no such a thing as a "complex" expression in a grouping set. Previously they'd fail during parsing. Now they fail during analysis with a slightly more descriptive error. This error in particular will only trigger for grouping sets containing complex expressions and the error will point to the right location, so it's not inaccurate. It's saying "group by expression you wrote at this location must be a column reference". It doesn't imply that all group by expressions must be column references.

**Comment by martint** [comment]
> I agree with @findepi that definitely for the case of simplegroupby it would be more accurate to call these groupingKeys.

Yeah, this one is a bit weird. It's due to the duality between simple group by expressions that allow complex expressions (in a departure from the spec) and grouping sets, which only allow column references. `groupingKeys` is also not accurate because from the point of view of the syntax, these just describe the set of columns in the grouping set specification. Let me think about a better name.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java [hunk 21]
```diff
@@ -131,4 +124,5 @@ static long calculateGrouping(long groupId, List<Integer> columns, List<Set<Inte
 
         return grouping;
     }
+
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java]: extra blank line
